### DISCUSSION TITLE
Zeek Docker restart policy leaves nodes in crashed or stopped state

### DISCUSSION THREAD

**Discussion by samanosuke26** [zeek/zeek#3004] [Help]
Problem as stated in the title
Zeek: 4.0.7
Host OS: CentOS 7
Docker version: 20.10.5

I have several hundred sensors and Zeek crashes are quite common. While I look into the source of the crashes, I want to automate the container restarting to recover from the crash. But when I test a restart policy and send a signal to crash a node (say, worker-1-1) the restart policy will kick off, restart the container, but zeek just comes back with worker-1-1 crashed still.

Has anyone run into this same problem or come up with a solution?
I have an entrypoint file that runs a zeekctl deploy on startup:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#!/bin/bash

setcap cap_net_raw,cap_net_admin=eip /opt/zeek/bin/zeek
setcap cap_net_raw,cap_net_admin=eip /opt/zeek/bin/capstats
runuser zeek -c '/opt/zeek/bin/zeekctl deploy'

trap "runuser zeek -c '/opt/zeek/bin/zeekctl stop'" SIGTERM
sleep infinity& wait; kill $!

**Comment by awelzel** [comment]
>  I want to automate the container restarting to recover from the crash. But when I test a restart policy and send a signal to crash a node (say, worker-1-1) the restart policy will kick off, restart the container, but zeek just comes back with worker-1-1 crashed still.

Could you share few details about
* How many Zeek workers do you run within an individual container?
* Details about the configuration of the restart policy and how you send/test the crash (are you sending signals to individual processes in the container?)
* Do you bind-mount any state/volumes into a container? If containers are stateless they shouldn't report crashed workers after a restart.


Taking a step back, you may be stumbling over "supervisor in a supervisor" hurdles here attempting to leverage Docker's container restart policy while at the same time using `zeekctl` within a container. Zeekctl usually launches multiple Zeek processes and offers `zeekctl cron` for restarting crashed processes.


---

This discussion may be better on Slack or the Forum, feel free to re-post continue there. You can join here:

Forum: https://community.zeek.org/
Slack: https://zeek.org/community/

**Reply by samanosuke26** [reply]
Thanks for the reply @awelzel 
Here are the details for your answers
_How many Zeek workers do you run within an individual container?_
**It depends on the traffic the device is ingesting. Anywhere from 1-10**

_Details about the configuration of the restart policy and how you send/test the crash (are you sending signals to individual processes in the container?)_
**I modified the running container with a restart policy of "always" to test it. I sent signals a few different ways.
docker exec zeek bash -c "kill 1"
docker kill --signal="SIGUSR1" zeek**
_Do you bind-mount any state/volumes into a container? If containers are stateless they shouldn't report crashed workers after a restart._
**Yes I bind-mount several volumes for logs and configurations**

I double checked the zeekctl cron configs and while it is enabled, there are currently no crons set. But that does give me another area I can try instead of the docker policy. I'll be looking into that, thank you.
